monai/auto3dseg/analyzer.py

```diff
@@ -0,0 +1,688 @@
+
+__all__ = [
+    "Analyzer",
+    "ImageStats",
+    "FgImageStats",
+    "LabelStats",
+    "ImageStatsSumm",
+    "FgImageStatsSumm",
+    "LabelStatsSumm",
+    "FilenameStats",
+]
+
+
+class Analyzer(MapTransform, ABC):
+    """
+    The Analyzer component is a base class. Other classes inherit this class will provide a callable
+    with the same class name and produces one pre-formatted dictionary for the input data. The format
+    is pre-defined by the init function of the class that inherit this base class. Function operations
+    can also be registerred before the runtime of the callable.
+
+    Args:
+        report_format: a dictionary that outlines the key structures of the report format.
+
+    """
+
+    def __init__(self, stats_name: str, report_format: dict) -> None:
+        super().__init__(None)
+        parser = ConfigParser(report_format)
+        self.report_format = parser.config
+        self.stats_name = stats_name
+        self.ops = ConfigParser({})
+
+    def update_ops(self, key: str, op: OperationType):
+        """
+        Register an statistical operation to the Analyzer and update the report_format
+
+        Args:
+            key: value key in the report.
+            op: Operation sub-class object that represents statistical operations.
+
+        """
+        self.ops[key] = op
+        parser = ConfigParser(self.report_format)
+
+        if parser.get(key, "NA") != "NA":
+            parser[key] = op
+
+        self.report_format = parser.config
+
+    def update_ops_nested_label(self, nested_key: str, op: OperationType):
+        """
+        Update operations for nested label format. Operation value in report_format will be resolved
+        to a dict with only keys
+
+        Args:
+            nested_key: str that has format of 'key1#0#key2'.
+            op: Operation sub-class object that represents statistical operations.
+        """
+        keys = nested_key.split(ID_SEP_KEY)
+        if len(keys) != 3:
+            raise ValueError("Nested_key input format is wrong. Please ensure it is like key1#0#key2")
+        root: str
+        child_key: str
+        (root, _, child_key) = keys
+        if root not in self.ops:
+            self.ops[root] = [{}]
+        self.ops[root][0].update({child_key: None})
+
+        self.ops[nested_key] = op
+
+        parser = ConfigParser(self.report_format)
+        if parser.get(nested_key, "NA") != "NA":
+            parser[nested_key] = op
+
+    def get_report_format(self):
+        """
+        Get the report format by resolving the registered operations recursively.
+
+        Returns:
+            a dictionary with {keys: None} pairs.
+
+        """
+        self.resolve_format(self.report_format)
+        return self.report_format
+
+    @staticmethod
+    def unwrap_ops(func: OperationType):
+        """
+        Unwrap a function value and generates the same set keys in a dict when the function is actually
+        called in runtime
```

Comment:
I think maybe we should describe the requirements for the `func`? like `data`, `data_addon`?

Thanks.

Reply:
Yes. I added some explanation as suggested.